lib/Dialect/LLHD/IR/LLHDOps.cpp

```diff
@@ -554,6 +615,55 @@ LogicalResult WaitOp::verify() {
 // HaltOp
 //===----------------------------------------------------------------------===//
 
+static LogicalResult deduplicateYieldOperands(PatternRewriter &rewriter,
+                                              HaltOp &op) {
+  auto processOp = dyn_cast<ProcessOp>(op->getParentOp());
+  if (!processOp || !processOp.getBody().hasOneBlock())
+    return failure();
+
+  auto yieldOperands = op.getYieldOperands();
+  llvm::SmallDenseMap<Value, unsigned> uniqueOperands;
+  llvm::SmallDenseMap<unsigned, unsigned> origToNewPos;
+  llvm::BitVector operandsToErase(yieldOperands.size());
+
+  for (auto [operandNo, operand] : llvm::enumerate(yieldOperands)) {
+    if (!uniqueOperands.contains(operand)) {
+      const auto newPos = uniqueOperands.size();
+      uniqueOperands.insert(std::make_pair(operand, newPos));
+      origToNewPos.insert(std::make_pair(operandNo, newPos));
+    } else {
+      auto firstOperandPos = uniqueOperands.lookup(operand);
+      origToNewPos.insert(std::make_pair(operandNo, firstOperandPos));
+      operandsToErase.set(operandNo);
+    }
+  }
+
+  if (uniqueOperands.size() == yieldOperands.size())
+    return failure();
+
+  op->eraseOperands(operandsToErase);
+  SmallVector<Type> resultTypes = llvm::to_vector(op->getOperandTypes());
+
+  rewriter.setInsertionPoint(processOp);
+  auto newProcessOp =
+      ProcessOp::create(rewriter, processOp->getLoc(), resultTypes,
+                        processOp->getOperands(), processOp->getAttrs());
+  newProcessOp.getBody().takeBody(processOp.getBody());
+
+  for (auto oldResult : processOp.getResults()) {
+    auto newResultPos = origToNewPos.lookup(oldResult.getResultNumber());
+    auto newResult = newProcessOp.getResult(newResultPos);
+    rewriter.replaceAllUsesWith(oldResult, newResult);
+  }
+
+  rewriter.eraseOp(processOp);
+  return success();
+}
+
+LogicalResult HaltOp::canonicalize(HaltOp op, PatternRewriter &rewriter) {
+  return deduplicateYieldOperands(rewriter, op);
+}
+
```

Comment:
This looks like it's doing the same thing as the canonicalizer for `ProcessOp` above 🤔. Since deduplicating yield operands requires us to change the process' results, which requires a new process op, it probably makes sense to do all of this as part of the `ProcessOp` canonicalizer above. WDYT?